Deblobbed
* drivers/gpu/drm/radeon/radeon-cp.c: Adjust for deblobbed sources.

diff -up linux-2.6.27.noarch/drivers/gpu/drm/radeon/radeon_cp.c.dma linux-2.6.27.noarch/drivers/gpu/drm/radeon/radeon_cp.c
--- linux-2.6.27.noarch/drivers/gpu/drm/radeon/radeon_cp.c.dma	2009-02-21 08:04:01.000000000 +1000
+++ linux-2.6.27.noarch/drivers/gpu/drm/radeon/radeon_cp.c	2009-02-21 08:07:08.000000000 +1000
@@ -726,9 +726,11 @@ static int radeon_do_engine_reset(struct
 }
 
 static void radeon_cp_init_ring_buffer(struct drm_device * dev,
-				       drm_radeon_private_t * dev_priv)
+				       drm_radeon_private_t * dev_priv,
+				       struct drm_file *file_priv)
 {
 	u32 ring_start, cur_read_ptr;
+	struct drm_radeon_master_private *master_priv;
 
 	/* Initialize the memory controller. With new memory map, the fb location
 	 * is not changed, it should have been properly initialized already. Part
@@ -859,6 +861,16 @@ static void radeon_cp_init_ring_buffer(s
 	dev_priv->scratch[6] = 0;
 	RADEON_WRITE(RADEON_SCRATCH_REG6, 0);
 
+	/* reset sarea copies of these */
+	if (file_priv) {
+		master_priv = file_priv->master->driver_priv;
+		if (master_priv->sarea_priv) {
+			master_priv->sarea_priv->last_frame = 0;
+			master_priv->sarea_priv->last_dispatch = 0;
+			master_priv->sarea_priv->last_clear = 0;
+		}
+	}
+
 	radeon_do_wait_for_idle(dev_priv);
 
 	/* Sync everything up */
@@ -1430,7 +1442,7 @@ static int radeon_do_init_cp(struct drm_
 	radeon_do_cleanup_cp(dev);
 	return -EINVAL;
 	/*(DEBLOBBED)*/
-	radeon_cp_init_ring_buffer(dev, dev_priv);
+	radeon_cp_init_ring_buffer(dev, dev_priv, file_priv);
 
 	dev_priv->last_buf = 0;
 
@@ -1498,7 +1510,7 @@ static int radeon_do_cleanup_cp(struct d
  *
  * Charl P. Botha <http://cpbotha.net>
  */
-static int radeon_do_resume_cp(struct drm_device * dev)
+static int radeon_do_resume_cp(struct drm_device * dev, struct drm_file *file_priv)
 {
 	drm_radeon_private_t *dev_priv = dev->dev_private;
 
@@ -1521,7 +1533,7 @@ static int radeon_do_resume_cp(struct dr
 	DRM_ERROR("Missing Free microcode!\n");
 	return -EINVAL;
 	/*(DEBLOBBED)*/
-	radeon_cp_init_ring_buffer(dev, dev_priv);
+	radeon_cp_init_ring_buffer(dev, dev_priv, file_priv);
 
 	radeon_do_engine_reset(dev);
 	radeon_irq_set_state(dev, RADEON_SW_INT_ENABLE, 1);
@@ -1721,7 +1733,7 @@ int radeon_cp_resume(struct drm_device *
 	if (drm_core_check_feature(dev, DRIVER_MODESET)) 
 		return 0;
 
-	return radeon_do_resume_cp(dev);
+	return radeon_do_resume_cp(dev, file_priv);
 }
 
 int radeon_engine_reset(struct drm_device *dev, void *data, struct drm_file *file_priv)
@@ -2419,7 +2431,7 @@ int radeon_modeset_cp_resume(struct drm_
 	DRM_ERROR("Missing Free microcode!\n");
 	return -EINVAL;
 	/*(DEBLOBBED)*/
-	radeon_cp_init_ring_buffer(dev, dev_priv);
+	radeon_cp_init_ring_buffer(dev, dev_priv, NULL);
 
 	radeon_do_engine_reset(dev);
 
